### PR TITLE
Do not complain about unused code when used in `impl` `Self` type

### LINKED ISSUE #18290: Spurious dead code warning when item is only used in impls
It seems that if a `const` is only used in `impl`s, but not in actual code, then rustc emits a "constant item is never used" warning, as in the following code:

``` Rust
const TLC: uint = 4;
trait Tr { fn doit(&self); }
impl Tr for [uint, ..TLC] {
    fn doit(&self) { println!("called 4"); }
}

fn main() {
    let s = [0,1,2,3u];
    s.doit(); // which .doit is called depends on architecture
}
```

Which prints (when compiled and then run):

```
<anon>:1:1: 1:21 warning: constant item is never used: `TLC`, #[warn(dead_code)] on by default
<anon>:1 const TLC: uint = 4;
         ^~~~~~~~~~~~~~~~~~~~
called 4
```

### PR DESCRIPTION
**Pull request by estebank**
Fix https://github.com/rust-lang/rust/issues/18290.

### CODE REVIEW

**Comment by estebank** [comment]
cc @arielb1 @Centril @Mark-Simulacrum

File: src/test/ui/lint/lint-dead-code-const-and-self.rs [hunk 1]
```diff
@@ -0,0 +1,35 @@
+// check-pass
+
+#![deny(dead_code)]
+
+const TLC: usize = 4;
```
Comment (estebank) [inline src/test/ui/lint/lint-dead-code-const-and-self.rs]: No complaint about TLC being unused

File: src/test/ui/lint/lint-dead-code-const-and-self.rs [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+trait Tr { fn doit(&self); }
+
+impl Tr for [usize; TLC] {
+    fn doit(&self) {
+        println!("called 4");
+    }
+}
+
+struct X;
+struct Y;
```
Comment (estebank) [inline src/test/ui/lint/lint-dead-code-const-and-self.rs]: No complaint about Y being unused

File: src/test/ui/derive-uninhabited-enum-38885.rs [hunk 3]
```diff
@@ -5,12 +5,15 @@
...
 #[derive(Debug)]
-enum Void {} //~ WARN never used
+enum Void {}
 
 #[derive(Debug)]
-enum Foo { //~ WARN never used
+enum Foo {
     Bar(u8),
-    Void(Void),
+    Void(Void), //~ WARN never used
```
Comment (estebank) [inline src/test/ui/derive-uninhabited-enum-38885.rs]: With this change we no longer complain about unused enums, only unused variants, which might not be what we want.

File: src/test/ui/derive-uninhabited-enum-38885.rs [hunk 4]
```diff
@@ -5,12 +5,15 @@
...
 }
 
-fn main() {}
+fn main() {
+    let x = Foo::Bar(42);
```
Comment (estebank) [inline src/test/ui/derive-uninhabited-enum-38885.rs]: This is added to remove the warning for `Foo::Bar`, not `Foo`.

File: src/test/ui/derive-uninhabited-enum-38885.rs [hunk 5]
```diff
@@ -5,12 +5,15 @@
 // when deriving Debug on an empty enum
 
 #[derive(Debug)]
-enum Void {} //~ WARN never used
+enum Void {}
```
Comment (estebank) [inline src/test/ui/derive-uninhabited-enum-38885.rs]: This test is to verify that _this_ doesn't emit a warning due to `derive`.

File: src/test/ui/derive-uninhabited-enum-38885.rs (hunk 3 above)
Comment (RalfJung) [reply src/test/ui/derive-uninhabited-enum-38885.rs]: > With this change we no longer complain about unused enums, only unused variants, which might not be what we want.

Seems like a regression to me. If an enum has no variants and is never used, it should still be warned about -- similar to a struct with no fields that is never used.

Comment (estebank) [reply src/test/ui/derive-uninhabited-enum-38885.rs]: The individual variants are warned about but the enum itself isn't. I believe it should be possible to have both, but is that something we want or is it just too verbose?

Comment (RalfJung) [reply src/test/ui/derive-uninhabited-enum-38885.rs]: If the enum has no variants, it's certainly not too verbose to warn at all at the enum itself.

**Comment by gagan0723** [comment]
Ping from triage, @matthewjasper can you please review this.

@estebank any updates after [RalfJung last comment](https://github.com/rust-lang/rust/pull/63317#discussion_r313093247).

Thanks.

File: src/test/ui/derive-uninhabited-enum-38885.rs (hunk 3 above)
Comment (estebank) [reply src/test/ui/derive-uninhabited-enum-38885.rs]: If the enum has no variants the current code complains about the enum. Added tests showing this and other permutations.

Comment (RalfJung) [reply src/test/ui/derive-uninhabited-enum-38885.rs]: Thanks!

**Review by matthewjasper — COMMENTED** [review]
r=me with comments addressed

**Pull request merged**